In [2]:
from src.biotools.fasta_tools import *
seq=ExtractSequence('P04529','../uvsx/data/curated_database/filtered_curated_database.fasta')

In [3]:
walker_a=seq[58:67]
walker_b=seq[137:143]

In [4]:
walker_b

'VVVFID'

In [69]:
import re


files = [
    '../uvsx/data/references/P04529.fasta',
    '../uvsx/data/references/9GBG.fasta',
    '../uvsx/data/references/7Z3M.fasta'
]

results = {}

for file in files:
    seq = read_fasta(file)
    name = file.split('/')[-1].split('.')[0]

    matches = []
    for match in re.finditer(r"[AVILMFWY]{4}DE?", seq):
        matches.append({
            "match": match.group(),
            "start": match.start(),
            "end": match.end()
        })

    results[name] = matches

print(results)


{'P04529': [{'match': 'VVFID', 'start': 138, 'end': 143}, {'match': 'LGAID', 'start': 338, 'end': 343}], '9GBG': [{'match': 'YGLLD', 'start': 276, 'end': 281}], '7Z3M': [{'match': 'VIGAD', 'start': 69, 'end': 74}, {'match': 'LFILD', 'start': 156, 'end': 161}]}


In [39]:
import pandas as pd
df=pd.read_csv('../uvsx/data/curated_database/cleaned_metadata.csv')


In [71]:
#Finding Walker A motif

a_results={}
a_present=[]
for acc in df['accession']:
    seq = ExtractSequence(acc, '../uvsx/data/curated_database/cleaned_curated_database.fasta')

    matches = []
    for match in re.finditer(r"[GA]....[GF]K[TS]", seq):
        matches.append({
            "match": match.group(),
            "start": match.start(),
            "end": match.end()
        })

    a_results[acc] = matches
    if len(matches)==0:
        print(f'{acc} WALKER A NOT FOUND')
        a_present.append(0)
    else:
        a_present.append(1)



Accession: P32270 Header: >sp|P32270.2|DDA_BPT4 RecName: Full=ATP-dependent DNA helicase dda; AltName: Full=DNA 5'-3' helicase Dda
Accession: P20315 Header: >sp|P20315.1|HELIC_BPT3 RecName: Full=DNA helicase/primase; AltName: Full=DNA 3'-5' helicase/primase Gp4; AltName: Full=Gene product 4; Short=Gp4
P20315 WALKER A NOT FOUND
Accession: UYE93698 Header: >UYE93698.1 DNA primase/helicase-like protein [Ralstonia phage 10RS305A]
UYE93698 WALKER A NOT FOUND
Accession: YDI96965 Header: >YDI96965.1 RecA protein [Staphylococcus phage SAP24]
YDI96965 WALKER A NOT FOUND
Accession: YDI96730 Header: >YDI96730.1 RecA protein [Staphylococcus phage SAP02]
YDI96730 WALKER A NOT FOUND
Accession: YDI94796 Header: >YDI94796.1 RecA-like DNA recombinase [Stenotrophomonas phage Solingen]
Accession: YDI94551 Header: >YDI94551.1 RecA-like DNA recombinase [Stenotrophomonas phage Siegen]
Accession: YDI93881 Header: >YDI93881.1 DNA binding protein uvsX [Pseudomonas phage Dresden]
YDI93881 WALKER A NOT FOUND
Acc

In [72]:
a_absent=[]

for acc in df['accession']:
    if len(a_results[acc]) == 0:
        a_absent.append(acc)
print(f'Number of Sequences without Walker A: {len(a_absent)}')

a_fs=[]
for acc in df['accession']:
    for match in a_results[acc]:
        seq=match.get('match')
        if seq[-3] == 'F':
            a_fs.append(acc)

print(f'Number of Sequences With F instead of G: {len(a_fs)}')

Number of Sequences without Walker A: 12325
Number of Sequences With F instead of G: 1464


In [73]:
# Finding walker B motif
b_results={}
b_present=[]
for acc in df['accession']:
    seq = ExtractSequence(acc, '../uvsx/data/curated_database/cleaned_curated_database.fasta')

    matches = []
    for match in re.finditer(r"[AVILMFWY]{4}DE?", seq):
        matches.append({
            "match": match.group(),
            "start": match.start(),
            "end": match.end()
        })

    b_results[acc] = matches
    if len(matches)==0:
        print(f'{acc} WALKER B NOT FOUND')
        b_present.append(0)
    else:
        b_present.append(1)

Accession: P32270 Header: >sp|P32270.2|DDA_BPT4 RecName: Full=ATP-dependent DNA helicase dda; AltName: Full=DNA 5'-3' helicase Dda
P32270 WALKER B NOT FOUND
Accession: P20315 Header: >sp|P20315.1|HELIC_BPT3 RecName: Full=DNA helicase/primase; AltName: Full=DNA 3'-5' helicase/primase Gp4; AltName: Full=Gene product 4; Short=Gp4
Accession: UYE93698 Header: >UYE93698.1 DNA primase/helicase-like protein [Ralstonia phage 10RS305A]
Accession: YDI96965 Header: >YDI96965.1 RecA protein [Staphylococcus phage SAP24]
YDI96965 WALKER B NOT FOUND
Accession: YDI96730 Header: >YDI96730.1 RecA protein [Staphylococcus phage SAP02]
YDI96730 WALKER B NOT FOUND
Accession: YDI94796 Header: >YDI94796.1 RecA-like DNA recombinase [Stenotrophomonas phage Solingen]
Accession: YDI94551 Header: >YDI94551.1 RecA-like DNA recombinase [Stenotrophomonas phage Siegen]
Accession: YDI93881 Header: >YDI93881.1 DNA binding protein uvsX [Pseudomonas phage Dresden]
Accession: YDI88709 Header: >YDI88709.1 RecA-like recombina

In [75]:
absent=[]
for acc in df['accession']:
    if len(results[acc]) == 0:
        absent.append(acc)
print(f'Number of Sequences without Walker B: {len(absent)}')



Number of Sequences without Walker B: 7542


In [77]:
hits=pd.read_csv('../uvsx/data/hmms/results/accession_hits.csv')

In [79]:
hits['walker_a']=a_present
hits['walker_b']=b_present

In [96]:
top_hits=[]
for index, row in hits.iterrows():
    if sum(row.drop('accession')) == 7:
        top_hits.append(row['accession'])

In [93]:
hits[hits['accession']=='YDI88709']

,accession,full_sequence_hmm_results,PF00154_custom_hmm_results,PF00154_hmm_results,PF21134_custom_hmm_results,PF21134_hmm_results,walker_a,walker_b
8,YDI88709,1,1,1,1,1,1,1


In [94]:
ExtractSequence('YDI88709', '../uvsx/data/curated_database/cleaned_curated_database.fasta')

Accession: YDI88709 Header: >YDI88709.1 RecA-like recombination protein [Escherichia phage phiM]


'MSDLKSRLIKASTSKMTADLTKSKLFNNRDEVPTRIPMLNIALGGALNAGLQSGLTIFAAPSKHFKTLFGLTMVAAYMKKYKDAICLFYDSEFGASESYFRSMGVDLDRVVHTPIQSVEQLKVDMTNQLDAIERGDKVIIFIDSIGNTASKKETEDALNEKVVGDMSRAKALKSLFRIVTPYLTIKDIPCVAINHTAMEIGGLYPKEIMGGGTGILYSANTVFFISKRQVKEGTELTGYDFTLKAEKSRTVKEKSTFPITVNFDGGIDPFSGLLEMATEIGFVVKPKAGWYAREFLDEETGEMIREEKSWRAKATDCVEFWGPLFKHKPFRDAIETKYKLGAISSIKEVDDAVNDLINCKATTKVPVKTSDAPSAADIENDLDEMEDFDE'

In [95]:
ExtractSequence('P04529', '../uvsx/data/curated_database/cleaned_curated_database.fasta')

Accession: P04529 Header: >sp|P04529.2|UVSX_BPT4 RecName: Full=Recombination and repair protein


'MSDLKSRLIKASTSKLTAELTASKFFNEKDVVRTKIPMMNIALSGEITGGMQSGLLILAGPSKSFKSNFGLTMVSSYMRQYPDAVCLFYDSEFGITPAYLRSMGVDPERVIHTPVQSLEQLRIDMVNQLDAIERGEKVVVFIDSLGNLASKKETEDALNEKVVSDMTRAKTMKSLFRIVTPYFSTKNIPCIAINHTYETQEMFSKTVMGGGTGPMYSADTVFIIGKRQIKDGSDLQGYQFVLNVEKSRTVKEKSKFFIDVKFDGGIDPYSGLLDMALELGFVVKPKNGWYAREFLDEETGEMIREEKSWRAKDTNCTTFWGPLFKHQPFRDAIKRAYQLGAIDSNEIVEAEVDELINSKVEKFKSPESKSKSAADLETDLEQLSDMEEFNE'

In [97]:
len(top_hits)

626

In [98]:
for acc in top_hits:
    seq=ExtractSequence(acc, '../uvsx/data/curated_database/cleaned_curated_database.fasta')
    if len(seq) < 300 and len(seq) > 400:
        print(f'{acc} has abbererant sequence\n{seq}')


Accession: YDI88709 Header: >YDI88709.1 RecA-like recombination protein [Escherichia phage phiM]
Accession: YDD38469 Header: >YDD38469.1 recombination protein [Escherichia phage vB_EcoM-DCSW23lw]
Accession: YCR04191 Header: >YCR04191.1 recombination and repair protein [Escherichia phage vB_EcoM_PhiSG_Ec1]
Accession: YCQ78089 Header: >YCQ78089.1 UvsX recombination and repair protein [Shewanella phage Persephone]
Accession: YCQ76618 Header: >YCQ76618.1 recombination and repair protein UvsX [Shewanella phage Hypnos]
Accession: P04529 Header: >sp|P04529.2|UVSX_BPT4 RecName: Full=Recombination and repair protein
Accession: YCL39723 Header: >YCL39723.1 recombination protein UvsX RecA [Aeromonas phage pAQF2A-18ST]
Accession: YCJ26573 Header: >YCJ26573.1 putative DNA recombination and repair protein RecA [Escherichia phage PG3]
Accession: YCJ26308 Header: >YCJ26308.1 putative DNA recombination and repair protein [Escherichia phage L73]
Accession: YCB26705 Header: >YCB26705.1 recombination and 

Okay so 626 sequences have a hit for all 7 checks. However with walker a and b these hits may be in sections of the protein that dont make sense
Potentially need to do filtering for motif hits that are in sketchy locations